internal/provider/data_source_consul_versions.go

```diff
@@ -0,0 +1,96 @@
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/hashicorp/terraform-provider-hcp/internal/clients"
+)
+
+// defaultConsulVersionsTimeoutDuration is the default timeout
+// for reading the agent Helm config.
+var defaultConsulVersionsTimeoutDuration = time.Minute * 5
+
+// dataSourceConsulVersions is the data source for the Consul versions supported by HCP.
+func dataSourceConsulVersions() *schema.Resource {
+	return &schema.Resource{
+		Description: "The Consul versions data source provides the Consul versions supported by HCP.",
+		Timeouts: &schema.ResourceTimeout{
+			Default: &defaultConsulVersionsTimeoutDuration,
+		},
+		ReadContext: dataSourceConsulVersionsRead,
+		Schema: map[string]*schema.Schema{
+			// Computed outputs
+			"recommended": {
+				Description: "The recommended Consul version for HCP clusters.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"available": {
+				Description: "The Consul versions available on HCP.",
+				Type:        schema.TypeList,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Computed: true,
+			},
+			"preview": {
+				Description: "The preview versions of Consul available on HCP.",
+				Type:        schema.TypeList,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Computed: true,
+			},
+		},
+	}
+}
+
+// dataSourceConsulVersionsRead is the func to implement reading of the
+// supported Consul versions on HCP.
+func dataSourceConsulVersionsRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	availableConsulVersions, err := clients.GetAvailableHCPConsulVersions(ctx, meta.(*clients.Client))
+	if err != nil || len(availableConsulVersions) == 0 {
+		return diag.Errorf("error fetching available HCP Consul versions: %v", err)
+	}
+
+	var recommendedVersion string
+	availableVersions := make([]string, 0)
+	previewVersions := make([]string, 0)
+
+	for _, v := range availableConsulVersions {
+		switch v.Status {
+		case consulmodels.HashicorpCloudConsul20200826VersionStatusRECOMMENDED:
+			recommendedVersion = v.Version
+			availableVersions = append(availableVersions, v.Version)
+		case consulmodels.HashicorpCloudConsul20200826VersionStatusAVAILABLE:
+			availableVersions = append(availableVersions, v.Version)
+		case consulmodels.HashicorpCloudConsul20200826VersionStatusPREVIEW:
+			previewVersions = append(previewVersions, v.Version)
+		}
+	}
+
+	err = d.Set("recommended", recommendedVersion)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	err = d.Set("available", availableVersions)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	err = d.Set("preview", previewVersions)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(fmt.Sprintf("recommended/%s/available_len/%d/preview_len/%d", recommendedVersion, len(availableVersions), len(previewVersions)))
```

Comment:
I don't think this would be blocking, but I'm curious if there are long-term plans for the id here. The intention seems to be to capture the version config at a specific point in time, but I think this format is technically vulnerable to things like 1 version added and 1 version removed, resulting in the same id. FWIW, I noticed AWS handles a similar use case by including a `syncToken` in their [ip-ranges](https://docs.aws.amazon.com/general/latest/gr/aws-ip-ranges.html) response.

Reply:
Good idea, I will go with a hash of the versions concated.